### PR TITLE
New Terms of Service Dialog

### PR DESCRIPTION
**Pull request by MrFlashAccount**
#### Tl;dr 
Closes: enso-org/cloud-v2#1228
This PR adds a new DIalog that requires user to submit terms and conditions

CleanShot 2024-05-16 at 16 44 52@2x

<details><summary>Demo Presentation</summary>
<p>

### **Note** during the demo the changes on the website were not merged yet, so when I clicked "view terms and conditions" link, I was redirected to the  / page. Once the changes on the website are merged, this link will lead to the "Terms and Conditions" page

https://github.com/enso-org/enso/assets/61194245/bd994319-9e9a-4227-a016-fede1136a956

</p>
</details> 

---

#### This Change:
What this change does in the larger context. Specific details to highlight for review:
1. Adds a new dialog that blocks the interactions with the dashboard until the user accepts terms and conditions
2. each 15 mins checks for changes in TOS policy(by fetching `eula.json` from the website with hash based on EULA content) and if the hash is different from the stored one, show the TOS dialog
3. Improves loading behavior (now 2 spinners are merged into the one)

#### Test Plan:
We expect to test the changes after merging https://github.com/enso-org/website/pull/7

---

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [ ] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [ ] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [ ] Unit tests have been written where possible.

### CODE REVIEW

File: app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    fetch(new URL('/eula.json', process.env.ENSO_CLOUD_ENSO_HOST))
+      .then(response => {
+        if (!response.ok) {
+          throw new Error('Failed to fetch terms of service')
+        } else {
+          return response.json()
+        }
+      })
+      .then(data => {
+        invariant(data != null, 'Invalid terms of service response')
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: @somebody1234 I think it's time to add a schema-validation library, WDYT?

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: we already have ajv as a dependency, is that not sufficient?

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: i guess ajv doesn't do type predicates...

File: app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        invariant(
+          // This is safe because we've checked that `data` is not null and that it has a `hash` property
+          // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+          'hash' in data && typeof data.hash === 'string',
+          'Invalid terms of service response'
+        )
+
+        // This is safe because we've checked the data above
+        // eslint-disable-next-line no-restricted-syntax
+        return data as { hash: string }
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: unnecessary type assertion. return `{ hash }` instead

File: app/ide-desktop/lib/dashboard/src/providers/LocalStorageProvider.tsx [hunk 3]
```diff
@@ -27,7 +27,14 @@ export interface LocalStorageProviderProps extends Readonly<React.PropsWithChild
 /** A React Provider that lets components get the shortcut registry. */
 export default function LocalStorageProvider(props: LocalStorageProviderProps) {
   const { children } = props
-  const localStorage = React.useMemo(() => new LocalStorage(), [])
+  const [, rerender] = React.useState<object>({})
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/providers/LocalStorageProvider.tsx]: please use the `useRefresh` hook (the purpose of `useState<object>({})` is not clear)

File: app/ide-desktop/lib/dashboard/src/providers/LocalStorageProvider.tsx [hunk 4]
```diff
@@ -27,7 +27,14 @@ export interface LocalStorageProviderProps extends Readonly<React.PropsWithChild
...
+  const localStorage = React.useMemo(
+    () =>
+      new LocalStorage(() => {
+        rerender({})
```
Comment (somebody1234) [inline app/ide-desktop/lib/dashboard/src/providers/LocalStorageProvider.tsx]: is there a reason we need to unconditionally rerender no matter which `LocalStorage` key is set?

**Comment by somebody1234** [comment]
CR: mostly fine apart from the points above?
i'm not sure about:
- firstly, requiring the user to go to another page to see the EULA
- secondly, putting the EULA *below* the confirmation checkbox

ideally we'd be showing the EULA in the app itself - I'm not sure whether an iframe works both in browser and in Electron, but if it is then that would be sufficient IMO (although we would probably want a special HTML page without the website header and footer, alongside the existing one

**Comment by PabloBuchu** [comment]
How do I QA this?

Screenshot 2024-05-17 at 11 19 30

**Comment by somebody1234** [comment]
@PabloBuchu see the PR description:
- We expect to test the changes after merging https://github.com/enso-org/website/pull/7

File: app/ide-desktop/lib/dashboard/src/providers/LocalStorageProvider.tsx (hunk 4 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/providers/LocalStorageProvider.tsx]: Yes, because we need to propagate the changes across components 
This's a hack though, eventually, we need to refactor `useLocalStorage` and allow `subscribing` to particular field in localStorage

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/providers/LocalStorageProvider.tsx]: right, just wondering whether it's necessary for every key - for example most of the other states (expanded columns, keyboard shortcuts) already set state as appropriate and hence they don't need a full rerender of the entire page.

although if it's been tested and there's no noticeable latency or anything when changing the other LocalStorage entries then there's no problem

File: app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx (hunk 1 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: I guess ajv should be sufficient

File: app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx (hunk 2 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: Unfortunately, ts complains that `hash` has `any` type. Maybe I'm doing something wrong?

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: sorry my bad, `{ hash: data.hash }`

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: CleanShot 2024-05-17 at 13 21 53@2x

Here's what I get 
Unfortunately typescript fails to infer the type

**Comment by MrFlashAccount** [comment]
@somebody1234 , @PabloBuchu just merged the changes on the website and I expect this PR to be ready for QA in a few minutes.

File: app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx (hunk 1 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: Uh oh... 
CleanShot 2024-05-17 at 14 36 36@2x

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: Do we need ajv though?

File: app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx (hunk 2 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: oh, that's because `any` breaks TS. see [TS playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBMAdgNwIZWKx8BmUIhwDkMSAngLRJoZYxEBQDAJgKYDGANuq3OxIgDO8ZqhioAXHACuiANaIIAd0RNq6TNgAUDOHrgwyYVhBxxR4uAF4bxCACMAVh3pwAZG-NjUcAIRW4RGlOTnc3XX0iAAtUQSiiBEQvSw8DIxMzC1QAOhi461siYQxEAHMiABoIvSIASRRUTmBmA1YoEEE4UzhBNuRgdl4oVkFIIVZGAEomfiF4JRj4AKyGAHpV-TgAPQB+BlnhOAWxACZrZJy8qLWN-V39gUPjmABmc4BvOCupLNzYqLgAF8bptdkA) for a working example

```ts
declare const data: unknown

invariant(
    typeof data === 'object' && data != null &&
    'hash' in data && typeof data.hash === 'string',
    'Invalid terms of service response'
)

const what = data
//    ^?
const what2 = data.hash
//    ^?
const what3 = { hash: data.hash }
//    ^?
```

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/TermsOfServiceModal.tsx]: Oh, awesome. I manually casted data to unknown

**Comment by MrFlashAccount** [comment]
PR is ready for review/QA

**Comment by PabloBuchu** [comment]
@MrFlashAccount truly minor things. When I hit `Accept` without checking the terms n cons. Can we make the checkbox and the text red? and maybe add `Please accept before continuing`

Also I think the checkbox should toggle when I click on text not only exactly in the square

**Comment by MrFlashAccount** [comment]
> Also I think the checkbox should toggle when I click on text not only exactly in the square

True, we're currently limited with the components we have. Eventually, we'll improve this. I moved this out of scope because this will unreasonably increase the scope of the PR.

**Comment by MrFlashAccount** [comment]
> When I hit Accept without checking the terms n cons. Can we make the checkbox and the text red? and maybe add Please accept before continuing


We currently focus on the checkbox but it is barely visible. Let's add an error message

**Comment by somebody1234** [comment]
> True, we're currently limited with the components we have

not sure this is true, a plain HTML `label` is more than enough to make this work. (similarly a react-aria-components `Checkbox` also provides this functionality out of the box.)

**Comment by MrFlashAccount** [comment]
> not sure this is true, a plain HTML label is more than enough to make this work. (similarly a react-aria-components Checkbox also provides this functionality out of the box.)

React aria expects the label to be passed as a child: https://react-spectrum.adobe.com/react-aria/Checkbox.html#example
But if I do so, our current checkbox implementation puts the label inside the checkmark box. 
I tried to connect the external label with the checkbox component but it seems like not working.

Reimplementing the checkbox will increase the scope of the feature and delay the release



<blockquote><div><strong><a href="https://react-spectrum.adobe.com/react-aria/Checkbox.html">Checkbox</a></strong></div><div>Documentation for Checkbox in the React Aria package.</div></blockquote>

**Comment by somebody1234** [comment]
> tried to connect the external label with the checkbox component

HTML `label` should work fine without any ids if you just contain both the checkbox and the text inside the `label`. but if that doesn't work then it's fine to leave it as-is for now.

**Comment by MrFlashAccount** [comment]
@somebody1234 can we approve the code and QA then?

**Review by somebody1234 — APPROVED** [review]
CR :white_check_mark: 
QA not done (it's late here)

**Review by PabloBuchu — APPROVED** [review]

**Comment by PabloBuchu** [comment]
ready to merge. please fix failing ci checks
